File: aoc_2022_javascript/day18.js
main();

function main() {
  const fs = require("fs");
  const input = fs
    .readFileSync("input/day18.txt", { encoding: "utf-8" })
    .trim()
    .split("\r\n");
  let start = Date.now();
  solution = solve(input);
  let end = Date.now();
  console.log(`Execution time: ${end - start} ms`);
  console.log(`Part 1: ${solution[0]}`);
  console.log(`Part 2: ${solution[1]}`);
}

function solve(lines) {
  let points = [];
  for (let line of lines) {
    points.push(line.split(",").map(x => parseInt(x)));
  }
  let maxDistance = 0;
  for (let point of points) {
    maxDistance = Math.max(maxDistance, Math.abs(point[0]));
    maxDistance = Math.max(maxDistance, Math.abs(point[1]));
    maxDistance = Math.max(maxDistance, Math.abs(point[2]));
  }
  let directions = [
    [1, 0, 0],
    [-1, 0, 0],
    [0, 1, 0],
    [0, -1, 0],
    [0, 0, 1],
    [0, 0, -1],
  ];
  let trappedCells = [];
  let part1 = 0;
  let part2 = 0;
  for (let point of points) {
    for (let dir of directions) {
      let cellFace = [point[0] + dir[0], point[1] + dir[1], point[2] + dir[2]];
      if (!contains(points, cellFace)) {
        part1++;
        let trapped = true;
        let openSet = [cellFace];
        let closedSet = [];
        while (openSet.length > 0) {
          currentCell = openSet.pop();
          closedSet.push(currentCell);
          if (contains(trappedCells, currentCell)) {
            break;
          }
          if (Math.abs(currentCell[0]) > maxDistance || Math.abs(currentCell[1]) > maxDistance || Math.abs(currentCell[2]) > maxDistance) {
            trapped = false;
            break;
          }
          for (let dirInner of directions) {
            let neighbor = [currentCell[0] + dirInner[0], currentCell[1] + dirInner[1], currentCell[2] + dirInner[2]];
            if (!contains(openSet, neighbor) && !contains(closedSet, neighbor) && !contains(points, neighbor)) {
              openSet.push(neighbor);
            }
          }
        }
        if (trapped) {
          for (let closed of closedSet) {
            if (!contains(trappedCells, closed)) {
              trappedCells.push(closed);
            }
          }
        } else {
          part2++;
        }
      }
    }
  }
  
  return [part1, part2];
}

function contains(array, element) {
  return array.filter(x => arrayEquals(x, element)).length > 0;
}

function arrayEquals(a, b) {
  return a[0] == b[0] && a[1] == b[1] && a[2] == b[2];
}

File: aoc_2024_python/day18.py
import aoc

def main(lines: list):
    part1 = None
    part2 = None

    grid_size = 70
    blockers = []
    for line in lines:
        a = line.split(",")
        blockers.append((int(a[0]), int(a[1])))
    
    path = dfs(blockers[:1024], grid_size)
    part1 = len(path)
    

    for nr_of_bytes in range(1024, len(lines)):
        if blockers[nr_of_bytes - 1] in path:
            path = dfs(blockers[:nr_of_bytes], grid_size)
        if not path:
            part2 = ",".join([str(a) for a in blockers[nr_of_bytes - 1]])
            break

    return part1, part2

def dfs(blockers, grid_size):
    directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]
    grid_range = range(grid_size + 1)
    
    open_set = [((0, 0), [])]
    closed_set = set()
    closed_set.add((0, 0))
    while open_set:
        pos, path = open_set.pop(0)
        if pos == (grid_size, grid_size):
            return path
        
        for dir in directions:
            pn = (pos[0] + dir[0], pos[1] + dir[1])
            if pn[0] in grid_range and pn[1] in grid_range and pn not in blockers and pn not in closed_set and pn not in open_set:
                open_set.append((pn, path + [pn]))
                closed_set.add(pn)
    return False

aoc.run_lines(main, "day18.txt")
